fdbserver/ClusterController.actor.cpp

```diff
@@ -3295,108 +3365,168 @@ void checkOutstandingStorageRequests(ClusterControllerData* self) {
-		     currentRKProcessId.present() && currentDDProcessId == currentRKProcessId &&
-		     (newRKWorker.interf.locality.processId() != self->masterProcessId &&
-		      newDDWorker.interf.locality.processId() != self->masterProcessId))) {
-			TraceEvent("CCHaltDD", self->id)
-			    .detail("DDID", db.distributor.get().id())
-			    .detail("Excluded", ddWorker.priorityInfo.isExcluded)
-			    .detail("Fitness", ddFitness)
-			    .detail("BestFitness", bestFitnessForDD)
-			    .detail("CurrentRateKeeperProcessId",
-			            currentRKProcessId.present() ? currentRKProcessId.get() : LiteralStringRef("None"))
-			    .detail("CurrentDDProcessId", currentDDProcessId)
-			    .detail("MasterProcessID", self->masterProcessId)
-			    .detail("NewRKWorkers", newRKWorker.interf.locality.processId())
-			    .detail("NewDDWorker", newDDWorker.interf.locality.processId());
-			ddWorker.haltDistributor = brokenPromiseToNever(
-			    db.distributor.get().haltDataDistributor.getReply(HaltDataDistributorRequest(self->id)));
+	auto rkSingleton = RatekeeperSingleton(db.ratekeeper);
+	auto ddSingleton = DataDistributorSingleton(db.distributor);
+
+	// Check if the singletons are healthy.
+	// side effect: try to rerecruit the singletons to more optimal processes
+	bool rkHealthy = isHealthySingleton<RatekeeperInterface>(
+	    self, newRKWorker, rkSingleton, bestFitnessForRK, self->recruitingRatekeeperID);
+
+	bool ddHealthy = isHealthySingleton<DataDistributorInterface>(
+	    self, newDDWorker, ddSingleton, bestFitnessForDD, self->recruitingDistributorID);
+
+	// if any of the singletons are unhealthy (rerecruited or not stable), then do not
+	// consider any further re-recruitments
+	if (!(rkHealthy && ddHealthy)) {
+		return;
+	}
+
+	// if we reach here, we know that the singletons are healthy so let's
+	// check if we can colocate the singletons in a more optimal way
+
+	// TODO: verify that we don't need to get the pid from the worker like we were doing before
+	Optional<Standalone<StringRef>> currentRKProcessId = rkSingleton.interface.get().locality.processId();
+	Optional<Standalone<StringRef>> currentDDProcessId = ddSingleton.interface.get().locality.processId();
+	Optional<Standalone<StringRef>> newRKProcessId = newRKWorker.interf.locality.processId();
+	Optional<Standalone<StringRef>> newDDProcessId = newRKWorker.interf.locality.processId();
+
+	auto currColocMap = getColocCounts({ currentRKProcessId, currentDDProcessId });
+	auto newColocMap = getColocCounts({ newRKProcessId, newDDProcessId });
+
+	auto currColocCounts = std::make_tuple(currColocMap[newRKProcessId], currColocMap[newDDProcessId]);
+	auto newColocCounts = std::make_tuple(newColocMap[newRKProcessId], newColocMap[newDDProcessId]);
+
+	// if the new coloc counts are collectively better (i.e. each singleton's coloc count has not increased)
+	if (newColocCounts <= currColocCounts) {
```

Comment:
I thought std::tuple ordering was lexicographical (i.e. compare first elements, if they're the same compare second elements). This logic wouldn't work for identifying that all tuple elements in newColocCounts are <= their corresponding element in currColocCounts. I think technically it works now since with 2 singletons they'll always have the same coloc count, but as soon as a new singleton type was added this wouldn't work.

Reply:
Ah yes - I didn't realize it was a lexicographical comparison. Fixing this by making the check explicit for each singleton.